File: crates/biorouter/src/agents/resource_refs.rs
//! Explicit resource references — the skills, extensions and knowledge bases a
//! user attaches to a message before sending it.
//!
//! # The canonical form is a tag
//!
//! ```text
//! <biorouter-ref type="skill" name="single-cell RNA &quot;QC&quot;">
//! <biorouter-ref type="extension" name="Chat Recall">
//! <biorouter-ref type="knowledge_base" id="soul" label="Soul &amp; Body">
//! ```
//!
//! Everything a composer emits uses this form. It exists because the compact
//! `/skill:name` markers cannot carry a name with a space in it: the extractor
//! for those splits the message on whitespace, so `/skill:my skill` reaches the
//! resolver as `my` and resolves to nothing (issue #65). Normalising the name
//! away — the fix that worked for `/ext:` in #60, where every consumer
//! re-normalises anyway — cannot transfer, because `skills_extension`'s
//! `loadSkill` looks the name up *exactly*: `myskill` is not `my skill`, so
//! that fix would trade a truncation for a quieter failure.
//!
//! The tag delimits the value explicitly instead of by whitespace, so any name
//! survives it.
//!
//! # Escaping — the contract other implementations must match
//!
//! Attribute values are escaped with a **fixed six-entry table** of XML
//! character references, applied to every occurrence regardless of position:
//!
//! | character | escaped as |
//! |-----------|------------|
//! | `&`       | `&amp;`    |
//! | `"`       | `&quot;`   |
//! | `<`       | `&lt;`     |
//! | `>`       | `&gt;`     |
//! | `\n`      | `&#10;`    |
//! | `\r`      | `&#13;`    |
//!
//! XML entities were chosen over percent-encoding or backslash escapes because
//! the syntax is already XML-shaped: `&quot;` is what a reader — human or
//! machine — expects to find inside `name="…"`, and it keeps the tag parseable
//! by a real XML parser should anyone ever want to use one.
//!
//! The table is **closed**. There is no general numeric-character-reference
//! support and no `&apos;`/`&nbsp;`: `&#65;` decodes to the literal text
//! `&#65;`, not to `A`. A closed table is a map an implementation in another
//! language can copy verbatim, which is the point — [`encode_ref_value`] and
//! [`decode_ref_value`] are exact inverses of each other and every emitter must
//! agree with them.
//!
//! Two rules make or break an implementation:
//!
//! * **Encoding must escape `&` first.** Written as a chain of replacements,
//!   `"<"` → `"&lt;"` → (`&` pass) → `"&amp;lt;"` corrupts every other escape.
//!   The implementation here iterates characters once and maps each
//!   independently, which is order-free by construction; a chained
//!   `String.replace` port must do `&` before all others.
//! * **Decoding must resolve `&amp;` last** for the mirror-image reason —
//!   otherwise a name containing the literal text `&quot;`, which encodes to
//!   `&amp;quot;`, decodes back to `"`. [`decode_ref_value`] scans left to
//!   right and consumes a whole entity at a time, which again removes the
//!   ordering question; a chained port must do `&amp;` last.
//!
//! `'` is deliberately *not* escaped, and single-quoted attribute values are
//! *not* accepted, so there is exactly one quoting style to agree on.
//!
//! # What a producer must guarantee
//!
//! A value is scanned to the next raw `"`, and a tag is abandoned at the next
//! raw `<`. Both are unreachable in correctly encoded output, so this costs
//! nothing — but a hand-written tag that skips the encoding will truncate.
//!
//! # Compatibility: the compact markers still parse
//!
//! `/skill:`, `/ext:`, `/kb:`, `/skill(…)`, `kb_id:` and the quoted legacy
//! phrases all still resolve. They appear in persisted sessions, saved
//! workflows, skill documents and the CLI completer, so removing them would
//! break messages already on disk. They keep their whitespace limitation — that
//! is now an acceptable tradeoff rather than a bug, because a producer that
//! needs to carry a name the compact form cannot represent has the tag to reach
//! for, and `reference_marker` picks between the two automatically.

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct KnowledgeBaseRef {
    pub id: String,
    pub label: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub(crate) struct ResourceRefs {
    pub skills: Vec<String>,
    pub extensions: Vec<String>,
    pub knowledge_bases: Vec<KnowledgeBaseRef>,
}

impl ResourceRefs {
    pub fn is_empty(&self) -> bool {
        self.skills.is_empty() && self.extensions.is_empty() && self.knowledge_bases.is_empty()
    }
}

/// The element name every reference tag uses.
pub const REF_TAG_NAME: &str = "biorouter-ref";

/// The escape table. Closed by design — see the module docs.
///
/// Order is irrelevant to both [`encode_ref_value`] and [`decode_ref_value`]:
/// each maps one character or one whole entity at a time rather than sweeping
/// the string once per rule, so neither can corrupt an escape it already
/// produced. No two entities share a first character after the `&`, so the
/// decoder's match is unambiguous.
const REF_ENTITIES: &[(char, &str)] = &[
    ('&', "&amp;"),
    ('"', "&quot;"),
    ('<', "&lt;"),
    ('>', "&gt;"),
    ('\n', "&#10;"),
    ('\r', "&#13;"),
];

/// Which kind of resource a reference names.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RefKind {
    Skill,
    Extension,
    KnowledgeBase,
}

impl RefKind {
    /// The `type="…"` keyword. A closed set, so it is never escaped.
    pub fn tag_type(self) -> &'static str {
        match self {
            RefKind::Skill => "skill",
            RefKind::Extension => "extension",
            RefKind::KnowledgeBase => "knowledge_base",
        }
    }

    /// The attribute the resource's identity travels in. Knowledge bases are
    /// named by `id` because that is what `kb_search` takes; skills and
    /// extensions by `name`.
    pub fn value_attr(self) -> &'static str {
        match self {
            RefKind::KnowledgeBase => "id",
            _ => "name",
        }
    }
}

/// Escape `value` for an attribute of a reference tag.
///
/// The inverse of [`decode_ref_value`]; see the module docs for the table and
/// for why the two must not be implemented as chained replacements.
pub fn encode_ref_value(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for ch in value.chars() {
        match REF_ENTITIES.iter().find(|(escaped, _)| *escaped == ch) {
            Some((_, entity)) => out.push_str(entity),
            None => out.push(ch),
        }
    }
    out
}

/// Unescape an attribute value from a reference tag.
///
/// Anything that is not one of the six entities in the table is passed through
/// untouched, including a lone `&`, an unknown entity (`&nbsp;`) and a numeric
/// reference outside the table (`&#65;`). Dropping them would silently mangle
/// names, and guessing at them would make this disagree with the encoder.
pub fn decode_ref_value(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    let mut rest = value;

    while let Some(amp) = rest.find('&') {
        out.push_str(&rest[..amp]);
        let tail = &rest[amp..];
        match REF_ENTITIES
            .iter()
            .find(|(_, entity)| tail.starts_with(*entity))
        {
            Some((decoded, entity)) => {
                out.push(*decoded);
                rest = &tail[entity.len()..];
            }
            // Not an entity we emit: keep the `&` verbatim and carry on from
            // the next character, so `&amp;` inside `&amp;amp;` still decodes
            // exactly once.
            None => {
                out.push('&');
                rest = &tail['&'.len_utf8()..];
            }
        }
    }

    out.push_str(rest);
    out
}

/// The canonical tag naming `value`.
pub fn ref_tag(kind: RefKind, value: &str) -> String {
    format!(
        "<{REF_TAG_NAME} type=\"{}\" {}=\"{}\">",
        kind.tag_type(),
        kind.value_attr(),
        encode_ref_value(value)
    )
}

/// The canonical tag naming `value`, carrying a display string for the chip a
/// UI renders in its place.
///
/// Only a knowledge base's label is read back out (it has somewhere to go — see
/// `KnowledgeBaseRef::label`); on the other kinds it is presentation-only and
/// the parser ignores it, which is exactly the tolerance a composer needs to
/// add attributes without coordinating with this module.
pub fn labelled_ref_tag(kind: RefKind, value: &str, label: &str) -> String {
    format!(
        "<{REF_TAG_NAME} type=\"{}\" {}=\"{}\" label=\"{}\">",
        kind.tag_type(),
        kind.value_attr(),
        encode_ref_value(value),
        encode_ref_value(label)
    )
}

pub(crate) fn extract_resource_refs(text: &str) -> ResourceRefs {
    let mut refs = ResourceRefs::default();

    extract_tag_refs(text, "skill", "name", &mut refs.skills);
    extract_tag_refs(text, "extension", "name", &mut refs.extensions);
    extract_kb_tag_refs(text, &mut refs.knowledge_bases);

    extract_legacy_resource_phrases(text, "skill", &mut refs.skills);
    extract_legacy_resource_phrases(text, "extension", &mut refs.extensions);
    extract_inline_refs(text, "skill:", &mut refs.skills);
    extract_inline_refs(text, "ext:", &mut refs.extensions);
    extract_inline_kb_refs(text, &mut refs.knowledge_bases);
    extract_function_refs(text, "/skill(", ")", &mut refs.skills);
    extract_function_refs(text, "/ext(", ")", &mut refs.extensions);
    extract_function_kb_refs(text, &mut refs.knowledge_bases);
    extract_legacy_kb_refs(text, &mut refs.knowledge_bases);

    dedup(&mut refs.skills);
    dedup(&mut refs.extensions);
    dedup_kbs(&mut refs.knowledge_bases);

    refs
}

// A `/ext:` reference is resolved to a concrete extension by
// `extension_manager::resolve_bundled_extension`, which keys off the extension
// id *and* the registry that owns it. This module only extracts the raw
// reference text from the message.

fn extract_tag_refs(text: &str, tag_type: &str, attr: &str, out: &mut Vec<String>) {
    let needle = format!("<biorouter-ref type=\"{tag_type}\" {attr}=\"");
    let mut rest = text;
    while let Some(start) = rest.find(&needle) {
        let value_start = start + needle.len();
        let value_rest = slice_from(rest, value_start);
        let Some(end) = value_rest.find('"') else {
            break;
        };
        push_trimmed(out, slice_to(value_rest, end));
        rest = slice_from(value_rest, end);
    }
}

fn extract_kb_tag_refs(text: &str, out: &mut Vec<KnowledgeBaseRef>) {
    let needle = "<biorouter-ref type=\"knowledge_base\" id=\"";
    let mut rest = text;
    while let Some(start) = rest.find(needle) {
        let id_start = start + needle.len();
        let id_rest = slice_from(rest, id_start);
        let Some(id_end) = id_rest.find('"') else {
            break;
        };
        let id = slice_to(id_rest, id_end).trim();
        if !id.is_empty() {
            out.push(KnowledgeBaseRef {
                id: id.to_string(),
                label: None,
            });
        }
        rest = slice_from(id_rest, id_end);
    }
}

fn extract_function_refs(text: &str, prefix: &str, suffix: &str, out: &mut Vec<String>) {
    let mut rest = text;
    while let Some(start) = rest.find(prefix) {
        let value_start = start + prefix.len();
        let value_rest = slice_from(rest, value_start);
        let Some(end) = value_rest.find(suffix) else {
            break;
        };
        push_trimmed(out, slice_to(value_rest, end));
        rest = slice_from(value_rest, end + suffix.len());
    }
}

fn extract_function_kb_refs(text: &str, out: &mut Vec<KnowledgeBaseRef>) {
    let mut refs = Vec::new();
    extract_function_refs(text, "/kb(", ")", &mut refs);
    for id in refs {
        out.push(KnowledgeBaseRef { id, label: None });
    }
}

fn extract_legacy_resource_phrases(text: &str, kind: &str, out: &mut Vec<String>) {
    let marker = format!("\" {kind}");
    let mut rest = text;
    while let Some(marker_start) = rest.find(&marker) {
        let before_marker = slice_to(rest, marker_start);
        let Some(open_quote) = before_marker.rfind('"') else {
            rest = slice_from(rest, marker_start + marker.len());
            continue;
        };
        push_trimmed(out, slice_from(before_marker, open_quote + 1));
        rest = slice_from(rest, marker_start + marker.len());
    }
}

fn extract_inline_refs(text: &str, prefix: &str, out: &mut Vec<String>) {
    for token in text.split_whitespace() {
        let trimmed = token.trim_matches(|c: char| c == ',' || c == '.' || c == ';');
        let value = trimmed.strip_prefix(&format!("/{prefix}"));
        if let Some(value) = value {
            push_trimmed(out, value);
        }
    }
}

fn extract_inline_kb_refs(text: &str, out: &mut Vec<KnowledgeBaseRef>) {
    for token in text.split_whitespace() {
        let trimmed = token.trim_matches(|c: char| c == ',' || c == '.' || c == ';');
        let value = trimmed.strip_prefix("/kb:");
        if let Some(value) = value {
            let id = value.trim();
            if !id.is_empty() {
                out.push(KnowledgeBaseRef {
                    id: id.to_string(),
                    label: None,
                });
            }
        }
    }
}

fn extract_legacy_kb_refs(text: &str, out: &mut Vec<KnowledgeBaseRef>) {
    let mut rest = text;
    while let Some(kb_start) = rest.find("kb_id:") {
        let after = slice_from(rest, kb_start + "kb_id:".len()).trim_start();
        let after = after.strip_prefix('"').unwrap_or(after);
        let after = after.strip_prefix('`').unwrap_or(after);
        let id_len = after
            .find(|c: char| c == ')' || c == '"' || c == '`' || c == ',' || c.is_whitespace())
            .unwrap_or(after.len());
        let id = slice_to(after, id_len).trim();
        if !id.is_empty() {
            out.push(KnowledgeBaseRef {
                id: id.to_string(),
                label: None,
            });
        }
        rest = slice_from(after, id_len);
    }

    let mut phrase_rest = text;
    while let Some(start) = phrase_rest.find("focus the \"") {
        let value_start = start + "focus the \"".len();
        let value_rest = slice_from(phrase_rest, value_start);
        let Some(end) = value_rest.find("\" knowledge base") else {
            break;
        };
        let id = slice_to(value_rest, end).trim();
        if !id.is_empty() {
            out.push(KnowledgeBaseRef {
                id: id.to_string(),
                label: None,
            });
        }
        phrase_rest = slice_from(value_rest, end);
    }
}

fn slice_from(value: &str, start: usize) -> &str {
    value.get(start..).unwrap_or("")
}

fn slice_to(value: &str, end: usize) -> &str {
    value.get(..end).unwrap_or(value)
}

fn push_trimmed(out: &mut Vec<String>, value: &str) {
    let value = value.trim();
    if !value.is_empty() {
        out.push(value.to_string());
    }
}

fn dedup(values: &mut Vec<String>) {
    let mut seen = std::collections::HashSet::new();
    values.retain(|value| seen.insert(value.clone()));
}

fn dedup_kbs(values: &mut Vec<KnowledgeBaseRef>) {
    let mut seen = std::collections::HashSet::new();
    values.retain(|value| seen.insert(value.id.clone()));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_machine_readable_refs() {
        let refs = extract_resource_refs(
            r#"<biorouter-ref type="skill" name="rna-qc"> <biorouter-ref type="extension" name="developer"> <biorouter-ref type="knowledge_base" id="soul">"#,
        );

        assert_eq!(refs.skills, vec!["rna-qc"]);
        assert_eq!(refs.extensions, vec!["developer"]);
        assert_eq!(
            refs.knowledge_bases,
            vec![KnowledgeBaseRef {
                id: "soul".to_string(),
                label: None
            }]
        );
    }

    #[test]
    fn extracts_legacy_inserted_phrases() {
        let refs = extract_resource_refs(
            r#"Use the "literature-review" skill for this request, Use the "pubmed" extension for this request, Using the Knowledge extension, focus knowledge base "Soul" (kb_id: soul) for this request"#,
        );

        assert_eq!(refs.skills, vec!["literature-review"]);
        assert_eq!(refs.extensions, vec!["pubmed"]);
        assert_eq!(refs.knowledge_bases[0].id, "soul");
    }

    #[test]
    fn extracts_inline_refs() {
        let refs = extract_resource_refs(
            "/skill:rna-qc /ext:developer /kb:soul /skill(rna-qc-v2) /ext(agentdrafter) /kb(project-notes)",
        );

        assert_eq!(refs.skills, vec!["rna-qc", "rna-qc-v2"]);
        assert_eq!(refs.extensions, vec!["developer", "agentdrafter"]);
        assert_eq!(refs.knowledge_bases[0].id, "soul");
        assert_eq!(refs.knowledge_bases[1].id, "project-notes");
    }

    // Alias coverage (`agentdrafter` -> `agent_drafter`, `autovisualizer` ->
    // `autovisualiser`) moved to
    // `extension_manager::tests::resolves_bundled_extension_spelling_aliases`,
    // which owns reference resolution now.

    /// The corpus every escaping test runs over: one entry per character class
    /// that has ever broken an escaping scheme, plus the two literal entity
    /// spellings that catch a wrong `&` ordering.
    const HOSTILE_NAMES: &[&str] = &[
        "rna-qc",
        "my skill",
        r#"say "hi""#,
        "tom & jerry",
        "a &amp; b",
        r#"a &quot; b"#,
        "&",
        "&&&",
        "<script>alert(1)</script>",
        "a < b > c",
        "commas, everywhere,",
        "line one\nline two",
        "carriage\r\nreturn",
        "naïve café — 生物路由器 🧬",
        r#"<biorouter-ref type="skill" name="nested">"#,
        "",
        "   ",
        "&#10;",
        "&#65;",
    ];

    /// The property the whole scheme rests on.
    #[test]
    fn escaping_round_trips_every_hostile_name() {
        for name in HOSTILE_NAMES {
            assert_eq!(
                decode_ref_value(&encode_ref_value(name)),
                *name,
                "round trip lost `{name}` (encoded as `{}`)",
                encode_ref_value(name)
            );
        }
    }

    /// The ampersand ordering, from both sides. A chained-replacement encoder
    /// that escapes `&` last produces `&amp;lt;` for `<`; a chained decoder
    /// that resolves `&amp;` first turns the encoding of the literal text
    /// `&quot;` into a bare `"`.
    #[test]
    fn the_ampersand_is_escaped_first_and_decoded_last() {
        assert_eq!(encode_ref_value("<"), "&lt;");
        assert_eq!(encode_ref_value("&"), "&amp;");
        assert_eq!(encode_ref_value("&lt;"), "&amp;lt;");
        assert_eq!(encode_ref_value(r#"&quot;"#), "&amp;quot;");

        assert_eq!(decode_ref_value("&amp;lt;"), "&lt;");
        assert_eq!(decode_ref_value("&amp;quot;"), r#"&quot;"#);
        assert_eq!(decode_ref_value("&amp;amp;"), "&amp;");
    }

    /// Position-independent: every occurrence is escaped, not just the first,
    /// and no character outside the table is touched.
    #[test]
    fn escaping_covers_the_whole_table_and_nothing_else() {
        assert_eq!(
            encode_ref_value("a&b\"c<d>e\nf\rg"),
            "a&amp;b&quot;c&lt;d&gt;e&#10;f&#13;g"
        );
        // `'`, tab and non-ASCII are deliberately left alone.
        assert_eq!(encode_ref_value("it's\ta — b"), "it's\ta — b");
    }

    /// An entity we do not emit must survive decoding verbatim rather than
    /// being dropped or guessed at.
    #[test]
    fn an_unknown_or_malformed_entity_is_left_alone() {
        for input in [
            "caf&eacute;",
            "&#65;",
            "&#x41;",
            "&apos;",
            "&nbsp;",
            "AT&T",
            "a & b",
            "&amp",
            "&",
            "&;",
            "&quot",
        ] {
            assert_eq!(decode_ref_value(input), input, "mangled `{input}`");
        }

        // ...and a known entity next to an unknown one still decodes.
        assert_eq!(decode_ref_value("&nbsp;&amp;&nbsp;"), "&nbsp;&&nbsp;");
    }

    #[test]
    fn tag_builders_escape_the_values_they_embed() {
        assert_eq!(
            ref_tag(RefKind::Skill, r#"say "hi" & bye"#),
            r#"<biorouter-ref type="skill" name="say &quot;hi&quot; &amp; bye">"#
        );
        assert_eq!(
            ref_tag(RefKind::Extension, "Chat Recall"),
            r#"<biorouter-ref type="extension" name="Chat Recall">"#
        );
        assert_eq!(
            labelled_ref_tag(RefKind::KnowledgeBase, "soul", "Soul & Body"),
            r#"<biorouter-ref type="knowledge_base" id="soul" label="Soul &amp; Body">"#
        );
    }
}

File: ui/desktop/src/utils/resourceRefs.ts
/**
 * The renderer's half of the `<biorouter-ref …>` contract (issue #65).
 *
 * A reference to a skill, extension or knowledge base travels in the message
 * text as a tag:
 *
 * ```text
 * <biorouter-ref type="skill" name="single-cell RNA &quot;QC&quot;">
 * <biorouter-ref type="extension" name="Chat Recall">
 * <biorouter-ref type="knowledge_base" id="soul" label="Soul &amp; Body">
 * ```
 *
 * The compact `/skill:name` markers it replaces cannot carry a name containing
 * whitespace: the backend extractor for those splits the message on whitespace,
 * so `/skill:my skill` reaches the resolver as `my`. Dropping the space instead
 * — the fix that worked for `/ext:` in issue #60, where every consumer
 * re-normalises anyway — cannot transfer, because `loadSkill` looks a skill up
 * *exactly*: `myskill` is not `my skill`.
 *
 * ## The escaping contract
 *
 * This module is a port. `crates/biorouter/src/agents/resource_refs.rs` is the
 * source of truth and its module docs are the spec; the two are pinned together
 * by `crates/biorouter/src/agents/resource_ref_corpus.json`, which both test
 * suites read. The table is closed — six entries, no others:
 *
 * | character | escaped as |
 * |-----------|------------|
 * | `&`       | `&amp;`    |
 * | `"`       | `&quot;`   |
 * | `<`       | `&lt;`     |
 * | `>`       | `&gt;`     |
 * | `\n`      | `&#10;`    |
 * | `\r`      | `&#13;`    |
 *
 * There is no general numeric-reference support and no `&apos;`/`&nbsp;`:
 * `&#65;` decodes to the literal text `&#65;`, not to `A`. `'` is never escaped
 * and single-quoted attribute values are rejected, so there is exactly one
 * quoting style for the two implementations to agree on.
 *
 * Both functions map one character (or one whole entity) at a time rather than
 * sweeping the string once per rule, which makes them order-free. Written as
 * chained `String.replace` calls instead, encoding must escape `&` **first**
 * (or `<` becomes `&amp;lt;`) and decoding must resolve `&amp;` **last** (or a
 * name containing the literal text `&quot;` decodes to a bare `"`).
 */

/** The element name every reference tag uses. */
export const REF_TAG_NAME = 'biorouter-ref';

/** Which kind of resource a reference names. Mirrors Rust's `RefKind`. */
export type RefKind = 'skill' | 'extension' | 'knowledge_base';

/**
 * The attribute a kind's identity travels in. Knowledge bases are named by `id`
 * because that is what `kb_search` takes; skills and extensions by `name`.
 */
export const refValueAttr = (kind: RefKind): 'id' | 'name' =>
  kind === 'knowledge_base' ? 'id' : 'name';

const REF_ENTITIES: ReadonlyArray<readonly [string, string]> = [
  ['&', '&amp;'],
  ['"', '&quot;'],
  ['<', '&lt;'],
  ['>', '&gt;'],
  ['\n', '&#10;'],
  ['\r', '&#13;'],
];

/** Escape `value` for an attribute of a reference tag. */
export function encodeRefValue(value: string): string {
  let out = '';
  for (const char of value) {
    const entity = REF_ENTITIES.find(([raw]) => raw === char);
    out += entity ? entity[1] : char;
  }
  return out;
}

/**
 * Unescape an attribute value from a reference tag.
 *
 * Anything outside the table survives untouched — a lone `&`, `&nbsp;`,
 * `&#65;`. Dropping them would mangle names and guessing at them would make
 * this disagree with {@link encodeRefValue}.
 */
export function decodeRefValue(value: string): string {
  let out = '';
  let index = 0;

  while (index < value.length) {
    if (value[index] !== '&') {
      out += value[index];
      index += 1;
      continue;
    }

    const entity = REF_ENTITIES.find(([, escaped]) => value.startsWith(escaped, index));
    if (entity) {
      out += entity[0];
      index += entity[1].length;
    } else {
      // Not an entity we emit: keep the `&` verbatim and resume one character
      // on, so `&amp;` inside `&amp;amp;` still decodes exactly once.
      out += '&';
      index += 1;
    }
  }

  return out;
}

/** The canonical tag naming `value`. */
export function refTag(kind: RefKind, value: string): string {
  return `<${REF_TAG_NAME} type="${kind}" ${refValueAttr(kind)}="${encodeRefValue(value)}">`;
}

/**
 * The canonical tag naming `value`, carrying a display string for the chip.
 *
 * Only a knowledge base's label is read back by the backend (it has somewhere
 * to go); on the other kinds it is presentation-only and the parser ignores it.
 * That tolerance is what lets the composer add chip attributes without
 * coordinating with the backend.
 */
export function labelledRefTag(kind: RefKind, value: string, label: string): string {
  return `<${REF_TAG_NAME} type="${kind}" ${refValueAttr(kind)}="${encodeRefValue(
    value
  )}" label="${encodeRefValue(label)}">`;
}

/** A reference the parser recognised, with the span of source it consumed. */
export interface RefSpan {
  kind: RefKind;
  /** The decoded identity: a skill/extension name, or a knowledge-base id. */
  value: string;
  /** The decoded display string, when the tag carried one. */
  label?: string;
  /** Byte offset of the `<` this tag starts at. */
  start: number;
  /** Byte offset one past the closing `>`. */
  end: number;
  /** The exact source the tag occupied, for a verbatim fallback. */
  raw: string;
}

const REF_KINDS: RefKind[] = ['skill', 'extension', 'knowledge_base'];

const isRefKind = (value: string): value is RefKind => (REF_KINDS as string[]).includes(value);

const leadingWhitespace = (input: string): number => input.length - input.trimStart().length;

/**
 * Parse the attribute list following the tag name, up to and including the `>`.
 *
 * Deliberately permissive about shape — any attribute order, extra attributes,
 * a valueless attribute, `/>` or `>`, whitespace anywhere — because the
 * alternative is losing a reference the user explicitly attached over a
 * cosmetic difference in serialisation. Strict about exactly two things: the
 * value's quoting (double quotes, entity-escaped) and that the tag is closed.
 *
 * Returns `null` when the tag never closes. An unterminated tag is dropped
 * rather than half-honoured: a message truncated mid-attribute must not load a
 * resource nobody named.
 */
function parseTagAttrs(input: string): { attrs: [string, string][]; consumed: number } | null {
  const attrs: [string, string][] = [];
  let pos = 0;

  for (;;) {
    if (pos > input.length) return null;
    pos += leadingWhitespace(input.slice(pos));
    const rest = input.slice(pos);

    if (rest.startsWith('/>')) return { attrs, consumed: pos + 2 };
    if (rest.startsWith('>')) return { attrs, consumed: pos + 1 };
    if (rest === '') return null;

    const nameEnd = rest.search(/[\s=>/]/);
    const nameLen = nameEnd === -1 ? rest.length : nameEnd;
    // A stray `=` or `/` where an attribute name belongs: not a shape we
    // understand, so leave the whole tag alone.
    if (nameLen === 0) return null;
    const name = rest.slice(0, nameLen);
    pos += nameLen;

    pos += leadingWhitespace(input.slice(pos));
    if (input[pos] !== '=') {
      // A valueless attribute (`<biorouter-ref … data-chip>`). Record it empty
      // and keep going rather than discarding an otherwise good tag over a
      // decoration.
      attrs.push([name, '']);
      continue;
    }
    pos += 1;
    pos += leadingWhitespace(input.slice(pos));

    // Double quotes only. Accepting `'` too would need `'` in the escape table
    // to be safe, and one quoting style is one fewer way for an emitter to
    // drift from this parser.
    if (input[pos] !== '"') return null;
    pos += 1;
    // Scanning to the next raw `"` is correct *because* of the escaping: an
    // escaped quote is `&quot;`, which contains no quote at all, so the first
    // one found is always the closing one.
    const end = input.indexOf('"', pos);
    if (end === -1) return null;
    attrs.push([name, input.slice(pos, end)]);
    pos = end + 1;
  }
}

/**
 * Every reference tag in `text`, in source order.
 *
 * Mirrors the backend's `parse_ref_tags` so the chip a user sees and the
 * resource the agent loads can never disagree: a tag rendered as a chip here is
 * one the agent resolves, and a tag left as text here is one the agent ignores.
 *
 * Tags are honoured wherever they appear, including inside a code fence — the
 * backend does the same, and a chip silently dropped because it sits in a fence
 * is worse than a fenced example being announced in the reply.
 */
export function findRefTags(text: string): RefSpan[] {
  const found: RefSpan[] = [];
  let index = 0;

  for (;;) {
    const start = text.indexOf('<', index);
    if (start === -1) break;
    // Advance past this `<` before anything else. Every failure below resumes
    // from here, which is what stops a malformed tag from either spinning
    // forever or eating the tags that follow it.
    index = start + 1;

    if (!text.startsWith(REF_TAG_NAME, index)) continue;
    const afterName = text.slice(index + REF_TAG_NAME.length);
    // `<biorouter-reference …>` is a different element: the name has to end
    // here, not merely start here.
    if (!/^[\s>/]/.test(afterName)) continue;

    // A tag never spans a raw `<`. A correctly encoded value carries `&lt;`
    // instead, so a raw one means this tag is malformed and another is starting.
    const nextTag = afterName.indexOf('<');
    const bounded = nextTag === -1 ? afterName : afterName.slice(0, nextTag);

    const parsed = parseTagAttrs(bounded);
    if (!parsed) continue;

    const attr = (key: string): string | undefined => {
      const hit = parsed.attrs.find(([name]) => name === key);
      // The raw value is trimmed *before* decoding, so slop in a hand-written
      // tag is forgiven while an encoded `&#10;` at either end survives.
      return hit ? hit[1].trim() : undefined;
    };

    const end = index + REF_TAG_NAME.length + parsed.consumed;
    index = end;

    // `type` is compared raw: a closed keyword set never needs escaping, and
    // decoding it would only invent a second spelling per type. An unknown type
    // is ignored rather than guessed at, so a newer composer can add one
    // without this build mis-filing it.
    const kind = attr('type') ?? '';
    if (!isRefKind(kind)) continue;

    const value = decodeRefValue(attr(refValueAttr(kind)) ?? '');
    if (value === '') continue;

    const label = attr('label') ? decodeRefValue(attr('label') as string) : undefined;

    found.push({
      kind,
      value,
      label: label || undefined,
      start,
      end,
      raw: text.slice(start, end),
    });
  }

  return found;
}

/** A run of plain text, or one reference, from {@link segmentRefTags}. */
export type RefSegment = { type: 'text'; text: string } | { type: 'ref'; ref: RefSpan };

/**
 * Split `text` into the runs of prose between its reference tags and the
 * references themselves, so a renderer can draw a chip in place of each tag.
 *
 * Text the parser did not claim comes back verbatim — including a malformed
 * tag, which the user then sees exactly as they typed it rather than as a blank
 * where a reference used to be.
 */
export function segmentRefTags(text: string): RefSegment[] {
  const segments: RefSegment[] = [];
  let cursor = 0;

  for (const ref of findRefTags(text)) {
    if (ref.start > cursor) segments.push({ type: 'text', text: text.slice(cursor, ref.start) });
    segments.push({ type: 'ref', ref });
    cursor = ref.end;
  }

  if (cursor < text.length) segments.push({ type: 'text', text: text.slice(cursor) });
  return segments;
}
